src/adapter/src/coord/caught_up.rs

```diff
@@ -385,4 +419,139 @@ impl Coordinator {
+
+        let mut replica_problem_counts: BTreeMap<ReplicaId, u32> = BTreeMap::new();
+
+        while let Some((source_data, _ts, diff)) = replica_statuses_stream.next().await {
+            // Only process inserts (positive diffs)
+            if diff <= 0 {
+                continue;
+            }
+
+            // Extract the Row from SourceData
+            let row = match source_data.0 {
+                Ok(row) => row,
+                Err(_) => continue, // Skip rows with errors
+            };
+
+            let mut iter = row.into_iter();
+
+            let replica_id: ReplicaId = iter
+                .next()
+                .expect("missing replica_id")
+                .unwrap_str()
+                .parse()
+                .expect("must parse as replica ID");
+            let _process_id = iter.next().expect("missing process_id").unwrap_uint64();
+            let status = iter
+                .next()
+                .expect("missing status")
+                .unwrap_str()
+                .to_string();
+            let reason_datum = iter.next().expect("missing reason");
+            let reason = if reason_datum.is_null() {
+                None
+            } else {
+                Some(reason_datum.unwrap_str().to_string())
+            };
+            let occurred_at = iter
+                .next()
+                .expect("missing occurred_at")
+                .unwrap_timestamptz();
+
+            // Only consider events within the time window and that are problematic
+            if occurred_at.naive_utc() >= min_timestamp_dt.naive_utc() {
+                if Self::is_problematic_status(&status, reason.as_deref()) {
+                    *replica_problem_counts.entry(replica_id).or_insert(0) += 1;
+                }
+            }
+        }
+
+        // Filter to replicas with 3 or more problematic events.
+        let result = replica_problem_counts
+            .into_iter()
+            .filter_map(|(replica_id, count)| {
+                if count >= 3 {
+                    tracing::info!(
+                        "Detected problematic cluster replica {}: {} problematic events in last {:?}",
+                        replica_id,
+                        count,
+                        Duration::from_millis(lookback_window)
+                    );
+                    Some(replica_id)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // Explicitly keep the read hold alive until this point.
+        drop(read_hold);
+
+        result
+    }
+
+    /// Determines if a replica status indicates a problematic state that could
+    /// indicate looping.
+    fn is_problematic_status(_status: &str, reason: Option<&str>) -> bool {
+        // For now, we only look at the reason, but we could change/expand this
+        // if/when needed.
+        if let Some(reason) = reason {
+            return reason == "oom-killed";
```

Comment:
We could factor out this string into a constant, and use it also at https://github.com/MaterializeInc/materialize/blob/563ac22eaaa8318eb38ad2fe396f32855d62c7b8/src/orchestrator/src/lib.rs#L108

Reply:
I'll change the code here to directly use `OfflineReason::OomKilled`, okay?

Reply:
Sounds good!